types/mongodb/index.d.ts

```diff
@@ -2378,12 +2429,82 @@ export class ChangeStream extends Readable {
 
 export class ResumeToken {}
 
+export type ChangeEventTypes = 'insert' | 'delete' | 'replace' | 'update' | 'drop' | 'rename' | 'dropDatabase' | 'invalidate';
+export interface ChangeEventBase<TSchema extends object = {_id: ObjectId}> {
+    _id: ResumeToken;
+    /**
+     * We leave this off the base type so that we can differentiate
+     * by checking its value and get intelligent types on the other fields
+     */
+    // operationType: ChangeEventTypes;
+    ns: {
+        db: string;
+        coll: string;
+    };
+    clusterTime: Timestamp;
+    txnNumber?: number;
+    lsid?: {
+        "id": any;
+        "uid": any;
+    };
+}
+export interface ChangeEventCR<TSchema extends object = {_id: ObjectId}> extends ChangeEventBase<TSchema> {
+    operationType: 'insert' | 'replace';
+    fullDocument?: TSchema;
+    documentKey: {
+        _id: TSchema extends {_id: any} ? TSchema['_id'] : any;
```

Comment:
We have a type `ExtractIdType` for extracting the `_id` type, can you use that one here?

```suggestion
        _id: ExtractIdType<TSchema>;
```

Reply:
Also in the other `documentKey`.

Reply:
I could, but my method is actually significantly more efficient; should I change ExtractIdType to use my method? The more steps used in these types the more likely you are to hit a max number of steps and error out the typescript compilation.

Reply:
I've replayed also there. I'm not sure it would be significant the difference. `ExtractIdType` is not overly complex

In any case it should be `TSchema extends {_id: any} ? TSchema['_id'] : ObjectId;` if we decide not to use `ExtractIdType`